core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java

```diff
@@ -34,72 +33,93 @@ private UfsSyncUtils() {} // prevent instantiation
+        syncPlan.setUpdateDirectory();
+      }
+      syncPlan.setSyncChildren();
+      return syncPlan;
+    }
+
+    // One of metadata or content is not in sync and it is a file
+    if (!isContentSynced) {
+      // update inode, by deleting and then optionally loading metadata
+      syncPlan.setDelete();
+      if (ufsExists) {
+        // UFS exists, so load metadata later.
+        syncPlan.setLoadMetadata();
+      }
+    } else {
+      syncPlan.setUpdateMetadata();
     }
     return syncPlan;
   }
 
   /**
-   * Returns true if the given inode is synced with the ufs status. This is a single inode check,
-   * so for directory inodes, this does not consider the children inodes.
+   * Returns true if the given inode's content is synced with the ufs status. This is a single inode
+   * check, so for directory inodes, this does not consider the children inodes.
    *
    * @param inode the inode to check for sync
    * @param ufsFingerprint the ufs fingerprint to check for the sync
    * @return true of the inode is synced with the ufs status
    */
-  public static boolean inodeUfsIsSynced(Inode inode, String ufsFingerprint) {
+  public static boolean inodeUfsIsContentSynced(Inode inode, Fingerprint ufsFingerprint) {
     boolean isSyncedUnpersisted =
-        !inode.isPersisted() && Constants.INVALID_UFS_FINGERPRINT.equals(ufsFingerprint);
-
+        !inode.isPersisted() && !ufsFingerprint.isValid();
     boolean isSyncedPersisted;
+
     if (inode instanceof InodeFile) {
       // check the file fingerprint.
       InodeFile inodeFile = (InodeFile) inode;
+      Fingerprint cachedFingerprint =  Fingerprint.parse(inodeFile.getUfsFingerprint());
       isSyncedPersisted = inodeFile.isPersisted()
-          && inodeFile.getUfsFingerprint().equals(ufsFingerprint)
-          && !inodeFile.getUfsFingerprint().equals(Constants.INVALID_UFS_FINGERPRINT);
+          && cachedFingerprint.matchContent(ufsFingerprint)
+          && cachedFingerprint.isValid();
     } else {
-      isSyncedPersisted = inode.isPersisted() && inode.getUfsFingerprint().equals(ufsFingerprint);
+      Fingerprint cachedFingerprint =  Fingerprint.parse(inode.getUfsFingerprint());
+      isSyncedPersisted = inode.isPersisted() && cachedFingerprint.matchContent(ufsFingerprint);
     }
     return isSyncedPersisted || isSyncedUnpersisted;
   }
 
+  /**
+   * Returns true if the given inode's metadata matches the ufs fingerprint.
+   *
+   * @param inode the inode to check for sync
+   * @param ufsFingerprint the ufs fingerprint to check for the sync
+   * @return true of the inode is synced with the ufs status
+   */
+  public static boolean inodeUfsIsMetadataSynced(Inode inode, Fingerprint ufsFingerprint) {
+    Fingerprint cachedFingerprint =  Fingerprint.parse(inode.getUfsFingerprint());
+    return cachedFingerprint.matchMetadata(ufsFingerprint);
+  }
+
   /**
    * A class describing how to sync an inode with the ufs.
    * A sync plan has several steps:
```

Comment:
Can you update these steps with the new version of the steps?

Reply:
done